lab_2_tokenize_by_bpe/main.py

```diff
@@ -47,7 +79,19 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
-
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    merge_token_dictionary = {}
+    for token in word_frequencies:
+        token_list = list(token)
+        if str(pair)[0 : 2] in str(token):
+            for index in range(len(token_list) - 1):
+                if (token[index], token[index + 1]) == pair:
+                    token_list[index: index + 2] = [''.join(pair)]
+            merge_token_dictionary[tuple(token_list)] = word_frequencies[token]
+        else:
+            merge_token_dictionary[token] = word_frequencies[token]
+    return merge_token_dictionary
```

Comment:
Can you explain the logic of this whole function?

Reply:
1) I go through all the dictionary keys and make a list of them.
2)If there is the right pair in this list, I go through the indexes of the elements
3)If token ([index], token[index+1]) = pair, I am replacing the two original elements with ' '.join(pair)
4) If pair is in token, I append tuple(token_list) as a key and word_frequencies[token] as value. If pair isn't in token, key is token